enip_backend/calls_gsheet_sync/run.py

```diff
@@ -0,0 +1,91 @@
+import logging
+
+import psycopg2
+from pytz import timezone
+
+from ..enip_common.config import CALLS_GSHEET_ID, POSTGRES_URL
+from ..enip_common.gsheets import get_gsheets_client, get_worksheet_data
+
+DATA_RANGE = ("A", "D")
+EXPECTED_DATA_HEADER = ["State", "AP Call", "AP Called At (ET)", "Published?"]
+
+PUBLISH_DEFAULT_RANGE = ("F1", "G1")
+
+CALLED_AT_TZ = timezone("US/Eastern")
+CALLED_AT_FMT = "%m/%d/%Y %H:%M:%S"
+
+
+def _update_calls_sheet_from_db(cursor, table, worksheet):
+    cursor.execute(
+        f"SELECT state, ap_call, ap_called_at FROM {table} WHERE ap_call IS NOT NULL"
+    )
+    db_calls_lookup = {r[0]: {"call": r[1], "called_at": r[2]} for r in cursor}
+    sheet_data = get_worksheet_data(worksheet, DATA_RANGE, EXPECTED_DATA_HEADER)
+    # Selectively update cells that don't have the same api calls from the db
+    # Note: we intentionally don't insert new rows into the sheet for states
+    # that exist in the db but not the sheet. The sheet will need to be seeded
+    # with the states we want to report on.
+    for row in sheet_data:
+        state = row["State"].value
+        db_call = db_calls_lookup.get(state)
+        if db_call:
+            call = db_call["call"]
+            ap_call_cell = row["AP Call"]
+            if ap_call_cell.value != call:
+                logging.info(f"Updating AP Call {state}: {call}")
+                ap_call_cell.set_value(call)
+
+            called_at_fmt = (
+                db_call["called_at"].astimezone(CALLED_AT_TZ).strftime(CALLED_AT_FMT)
+            )
+            ap_called_at_cell = row["AP Called At (ET)"]
+            if ap_called_at_cell.value != called_at_fmt:
+                logging.info(f"Updating AP Called At {state}: {called_at_fmt}")
+                ap_called_at_cell.set_value(called_at_fmt)
+
+
+def _update_db_published_from_sheet(cursor, table, worksheet):
+    publish_default_cells = worksheet.get_values(*PUBLISH_DEFAULT_RANGE)[0]
+    if publish_default_cells[0] != "Default":
+        raise Exception(
+            "Worksheet {} does not match the expected format".format(worksheet.title)
+        )
+    default_is_publish = publish_default_cells[1] == "Publish"
+
+    def handle_published(value):
+        if value == "Default":
+            return default_is_publish
+        return value == "Yes"
+
+    sheet_data = get_worksheet_data(worksheet, DATA_RANGE, EXPECTED_DATA_HEADER)
+    rows = [
+        (row["State"].value, handle_published(row["Published?"].value))
+        for row in sheet_data
+    ]
+    # Note: published settings are updated for all states in the sheet, even those that don't have calls
+    statement = f"""
+    INSERT INTO {table} (state, published)
+    VALUES (%s, %s)
+    ON CONFLICT (state) DO UPDATE
+    SET published = EXCLUDED.published
```

Comment:
Wondering if we need/Charlotte wants a mechanism by which we can call a race before the AP? just pinged the slack about it